src/app/ReadHandler.h

```diff
@@ -247,20 +289,33 @@ class ReadHandler : public Messaging::ExchangeDelegate
      * Notify the read handler that a set of attribute paths has been marked dirty.
      */
     void SetDirty(const AttributePathParams & aAttributeChanged);
-    bool IsDirty() const { return (mDirtyGeneration > mPreviousReportsBeginGeneration) || mForceDirty; }
-    void ClearDirty() { mForceDirty = false; }
+    bool IsDirty() const
+    {
+        return (mDirtyGeneration > mPreviousReportsBeginGeneration) || mFlags.Has(ReadHandlerFlags::ForceDirty);
+    }
+    void ClearDirty() { mFlags.Clear(ReadHandlerFlags::ForceDirty); }
 
-    NodeId GetInitiatorNodeId() const { return mInitiatorNodeId; }
-    FabricIndex GetAccessingFabricIndex() const { return mSubjectDescriptor.fabricIndex; }
+    NodeId GetInitiatorNodeId() const
+    {
+        auto session = GetSession();
+        return session == nullptr ? kUndefinedNodeId : session->GetPeerNodeId();
+    }
 
-    const SubjectDescriptor & GetSubjectDescriptor() const { return mSubjectDescriptor; }
+    FabricIndex GetAccessingFabricIndex() const
+    {
+        auto session = GetSession();
+        return session == nullptr ? kUndefinedFabricIndex : session->GetFabricIndex();
```

Comment:
Why are we OK assuming session is not null in GetSubjectDescriptor() but not here or in GetInitiatorNodeId?

Also, I am not clear on exactly how this is meant to work in terms of resource allocation and limits: if GetSession() is null, this read handler will not be treated as belonging to any fabric.  Should we make sure a ReadHandler always gets shut down if its underlying session goes away?

Ah, I see you have a TODO about that below... Yes, we should do it.